File: python/leetcode129.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        nums = []
        self.findnums(root, 0, nums)
        return sum(nums)

    def findnums(self,root, num, nums):
        if not root:
            return
        num = num*10+root.val
        if root.left == None and root.right == None:
            nums.append(num)
        self.findnums(root.left, num, nums)
        self.findnums(root.right, num, nums)

File: leetcodec++/leetcode129.cpp
//
// Created by 黄敏辉 on 2019-08-07.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        vector<int> nums;
        findNums(nums, root);
        int res = 0;
        for (int i = 0; i < nums.size(); i++)
            res += nums[i];
        return res;
    }

    void findNums(vector<int>& nums, TreeNode* root){
        if (!root)
            return;
        if (!root->left && !root->right){
            nums.push_back(root->val);
            return;
        }
        if (root->left){
            root->left->val = root->val * 10 + root->left->val;
        }
        findNums(nums, root->left);
        if (root->right){
            root->right->val = root->val * 10 + root->right->val;
        }
        findNums(nums, root->right);

        return;
    }
};